### ISSUE TITLE
序列标注模型运行出错

### ISSUE THREAD

**Issue by livingbody** [PaddlePaddle/PaddleNLP#654]
https://aistudio.baidu.com/aistudio/projectdetail/2050392
大佬们，我训练的模型，预测的时候，第一行第一列，也就是说test数据集第一句话总是掉一个字，要哭死了，但是我找不到出错原因。。。。。。。。

**Comment by joey12300** [comment]
你好，可以先看看输出的logits，len是否符合预期，再调试parse_decodes是否遗漏了部分token的解码

**Comment by livingbody** [comment]
发现只要带 “-” 的都少一个，奇怪，decodes看起来也淡淡无奇的样子。。。。。。
```
def parse_decodes(ds, decodes, lens, label_vocab):
    decodes = [x for batch in decodes for x in batch]
    lens = [x for batch in lens for x in batch]
    id_label = dict(zip(label_vocab.values(), label_vocab.keys()))
    tags=[]
    outputs = []
    for idx, end in enumerate(lens):
        sent = ds.data[idx][0][:end]
        tags = [id_label[x] for x in decodes[idx][1:end]]
        sent_out = []
        tags_out = []
        words = ""
        for s, t in zip(sent, tags):
            if t.startswith('B-'):
                if len(words):
                    sent_out.append(words)
                tags_out.append(t.split('-')[1])
                words = s
                tags_out.append(t.split('-')[1])
                words = s
            elif t == 'O':
                if len(words):
                    sent_out.append(words)
                tags_out.append(t.split('-')[0])
                words = s
            else:
                words += s
        if len(sent_out) < len(tags_out):
            sent_out.append(words)
        outputs.append(''.join(
            [str((s, t)) for s, t in zip(sent_out, tags_out)]))
    return outputs, tags
```

**Comment by livingbody** [comment]
···
def parse_decodes(ds, decodes, lens, label_vocab):
    decodes = [x for batch in decodes for x in batch]
    lens = [x for batch in lens for x in batch]
    id_label = dict(zip(label_vocab.values(), label_vocab.keys()))
    tags=[]
    outputs = []
    for idx, end in enumerate(lens):
        sent = ds.data[idx][0][:end]
        tags = [id_label[x] for x in decodes[idx][1:end]]
        sent_out = []
        tags_out = []
        words = ""
        for s, t in zip(sent, tags):
            if t.startswith('B-'):
                if len(words):
                    sent_out.append(words)
                tags_out.append(t.split('-')[1])
                words = s
                tags_out.append(t.split('-')[1])
                words = s
            elif t == 'O':
                if len(words):
                    sent_out.append(words)
                tags_out.append(t.split('-')[0])
                words = s
            else:
                words += s
        if len(sent_out) < len(tags_out):
            sent_out.append(words)
        outputs.append(''.join(
            [str((s, t)) for s, t in zip(sent_out, tags_out)]))
    return outputs, tags
···

**Comment by livingbody** [comment]
例如：
('朝阳区', 'district')('小关北里', 'district')('000-', 'road')('0', 'road')
('朝阳区', 'district')('惠新东街', 'district')('00', 'road')('号', 'road')
('朝阳区', 'district')('南磨房路', 'district')('与', 'road')('西大望路', 'road')('交口', 'O')('东南角', 'road')
第一条后面有个号，结果就少一个。。。。。。

**Comment by ZeyuChen** [comment]
@livingbody 后续还是建议issues标题使用具体的错误类型或者应用场景作为标题，否则这样的问题很难被其他开发者接受。

**Comment by livingbody** [comment]
后来认真检查，是遇到“-“这个符号，就会少一个字。。。。。。。

**Comment by livingbody** [comment]
但是不清楚为啥，代码看了好几遍

**Comment by livingbody** [comment]
> 你好，可以先看看输出的logits，len是否符合预期，再调试parse_decodes是否遗漏了部分token的解码

检查发现遇到”-“这个符号，就会出现丢字现象，是不是实体标注把”-“屏蔽了？

**Comment by livingbody** [comment]
全文搜索所使用的ernie-1.0中文模型所用到的https://paddlenlp.bj.bcebos.com/models/transformers/ernie/vocab.txt，里面没有“-”字符，是不是这个原因？
如果要用到该怎么做？手动改文件吗？

**Comment by livingbody** [comment]
> 全文搜索所使用的ernie-1.0中文模型所用到的[https://paddlenlp.bj.bcebos.com/models/transformers/ernie/vocab.txt，里面没有“-”字符，是不是这个原因？](https://paddlenlp.bj.bcebos.com/models/transformers/ernie/vocab.txt%EF%BC%8C%E9%87%8C%E9%9D%A2%E6%B2%A1%E6%9C%89%E2%80%9C-%E2%80%9D%E5%AD%97%E7%AC%A6%EF%BC%8C%E6%98%AF%E4%B8%8D%E6%98%AF%E8%BF%99%E4%B8%AA%E5%8E%9F%E5%9B%A0%EF%BC%9F)
> 如果要用到该怎么做？手动改文件吗？

跟战歌有关吗？

**Comment by livingbody** [comment]
> 全文搜索所使用的ernie-1.0中文模型所用到的[https://paddlenlp.bj.bcebos.com/models/transformers/ernie/vocab.txt，里面没有“-”字符，是不是这个原因？](https://paddlenlp.bj.bcebos.com/models/transformers/ernie/vocab.txt%EF%BC%8C%E9%87%8C%E9%9D%A2%E6%B2%A1%E6%9C%89%E2%80%9C-%E2%80%9D%E5%AD%97%E7%AC%A6%EF%BC%8C%E6%98%AF%E4%B8%8D%E6%98%AF%E8%BF%99%E4%B8%AA%E5%8E%9F%E5%9B%A0%EF%BC%9F)
> 如果要用到该怎么做？手动改文件吗？

和这个有关吗？

**Comment by libertatis** [comment]
这个问题出在测试集加载上，下面是你的测试集加载的代码：
```
def load_dataset(datafiles):
    def read(data_path):
        with open(data_path, 'r', encoding='utf-8') as fp:
            # next(fp)  # 没有header，不用Skip header
            for line in fp.readlines():
                ids, words = line.strip('\n').split('\001')
                # 要预测的数据集没有label，伪造个O，不知道可以不 ，应该后面预测不会用label
                labels=['O' for x in range(0,len(words))]
                yield words, labels

    if isinstance(datafiles, str):
        return MapDataset(list(read(datafiles)))
    elif isinstance(datafiles, list) or isinstance(datafiles, tuple):
        return [MapDataset(list(read(datafile))) for datafile in datafiles]  
```
你返回的 `words` 是一个字符串。而字符串直接交给 `tokenizer` 切分，它不一定是按照字符一个一个切分的。以测试集第一个样本为例：
```
from paddlenlp.transformers import ErnieTokenizer


text = '朝阳区小关北里000-0号'  # 这里 text 的长度为13
tokenizer = ErnieTokenizer.from_pretrained('ernie-1.0')
encoded = tokenizer(text=text)
print('input_ids: ', encoded['input_ids'])
print('len(input_ids): ', len(encoded['input_ids']))
```
输出 `encoded['input_ids']` 及其长度的结果为：
```
input_ids:  [1, 705, 507, 121, 96, 129, 252, 155, 2721, 17963, 540, 500, 2]
len(input_ids):  13
```
而 `input_ids` 包含了 `[CLS]` 和 `[SEP]` 的编码，所以 `text` 经过切词后，实际长度为 `11`。于是就出现了测试集句子长度和标签长度不一致的问题。
解决办法为，将测试集样本转化为字符列表，而非整个字符串交给 `tokenizer`去切分。
```
    words = [ch for ch in words]
    yield words
```

嗯，还有就是测试集不需要 `labels`，但是你可能需要那个 `id` ，尤其是当测试集中的 `id` 不是连续整数，而是一长串数字的时候，在将预测结果写入提交结果文件时需要~

**Comment by joey12300** [comment]
@livingbody @libertatis 
您好，对于纯中文而言，使用ErnieTokenizer可以按照字的粒度去切分，能满足正常的序列标注任务。而对于包含数字、英文的语料，则需要先使用特殊字符（如`\002`）去分割每个Token，然后再使用ErnieTokenizer去切分。

如“朝阳区小关北里000-0号”，可先用`\002`分割每个token，变为“朝\002阳\002区\002小\002关\002北\002里\0020\0020\0020\002-\0020\002号”，然后使用tokenizer切分。

序列标注任务可以参考MSAR NER任务：https://github.com/PaddlePaddle/PaddleNLP/tree/develop/examples/information_extraction/msra_ner
其中，有关tokenizer的用法可以参考：https://github.com/PaddlePaddle/PaddleNLP/blob/develop/examples/information_extraction/msra_ner/train.py#L74。

**Comment by libertatis** [comment]
> @livingbody @libertatis
> 您好，对于纯中文而言，使用ErnieTokenizer可以按照字的粒度去切分，能满足正常的序列标注任务。而对于包含数字、英文的语料，则需要先使用特殊字符（如`\002`）去分割每个Token，然后再使用ErnieTokenizer去切分。
> 
> 如“朝阳区小关北里000-0号”，可先用`\002`分割每个token，变为“朝\002阳\002区\002小\002关\002北\002里\0020\0020\0020\002-\0020\002号”，然后使用tokenizer切分。
> 
> 序列标注任务可以参考MSAR NER任务：https://github.com/PaddlePaddle/PaddleNLP/tree/develop/examples/information_extraction/msra_ner
> 其中，有关tokenizer的用法可以参考：[https://github.com/PaddlePaddle/PaddleNLP/blob/develop/examples/information_extraction/msra_ner/train.py#L74。](https://github.com/PaddlePaddle/PaddleNLP/blob/develop/examples/information_extraction/msra_ner/train.py#L74%E3%80%82)

老师您的意思是，需要先对测试集进行预处理吗？比如，先把测试集中的每个样本按字符 `\002` 分割，如：
```
text = “朝阳区小关北里000-0号” 
text = [ch + '\002' for ch in text]
text[-1] = text[-1][0]

with open(test.tsv, 'w', encoding='utf-8') as fWriter:
    fWriter.write(''.join([ch for ch in text]))

# “朝阳区小关北里000-0号” -> “朝\002阳\002区\002小\002关\002北\002里\0020\0020\0020\002-\0020\002号”
```
然后再重新保存到测试集文件中，加载数据集的时候，每个样本再按照 `\002` 分割：
```
def load_dataset(datafiles):
    def read(data_path):
        with open(data_path, 'r', encoding='utf-8') as fp:
            # next(fp)  # 没有header，不用Skip header
            for line in fp.readlines():
                ids, words = line.strip('\n').split('\001')
                # 要预测的数据集没有label，伪造个O，不知道可以不 ，应该后面预测不会用label
                labels=['O' for x in range(0,len(words))]
                
                # 按照 '\002' 分割
                words = words.split('\002')

                yield words, labels

    if isinstance(datafiles, str):
        return MapDataset(list(read(datafiles)))
    elif isinstance(datafiles, list) or isinstance(datafiles, tuple):
        return [MapDataset(list(read(datafile))) for datafile in datafiles]  
```
请问，老师您是这个意思吗？

**Comment by joey12300** [comment]
@libertatis 是的，切分完后还得用tokenizer处理一下。
```
words = words.split('\002')
tokenized_input = tokenizer(
        words,
        return_length=True,
        is_split_into_words=True,
        max_seq_len=max_seq_len)
```

**Comment by libertatis** [comment]
> @libertatis 是的，切分完后还得用tokenizer处理一下。
> 
> ```
> words = words.split('\002')
> tokenized_input = tokenizer(
>         words,
>         return_length=True,
>         is_split_into_words=True,
>         max_seq_len=max_seq_len)
> ```

那预先使用特殊字符 `\002` 分割每个 `token`，保存起来，加载数据集时，再对每个样本使用 `\002` 分割:
```
words = words.split('\002')
```
这样做的好处是什么呢？如果我并不对测试集预处理，而在加载数据集时，直接把样本字符串转化为一个字符列表：
```
words = [ch for ch in words]
```
这两种方式最终得到的 `words` 是一样的呀？也就是说送给 `tokenizer` 的输入是一样的呀？那为什么还要预先对样本使用 `\002` 分割预处理，这不是多此一举吗？嗯，这里面还有哪些是我没有考虑到的吗？烦请老师不吝赐教。谢谢啦~

**Comment by livingbody** [comment]
谢谢各位大佬，我实践实践先，了解了问题所在了。

**Comment by livingbody** [comment]
我在想各位老师是不打我一顿的心都有了

**Comment by livingbody** [comment]
各位大佬，比赛跑通了，接下来就是调优了。
